### ISSUE TITLE
Allow capture of imported modules to become a `require` at runtime

### ISSUE THREAD

**Issue by joeduffy** [pulumi/pulumi#1588]
**Note**: The originally reported issue below was addressed with https://github.com/pulumi/pulumi-aws/pull/262.  This issue is now tracking the proposals in https://github.com/pulumi/pulumi/issues/1588#issuecomment-401968174

In https://github.com/pulumi/pulumi-aws/issues/249, @rgwood reported that we had missed the `qs` and `uuid` modules imported within a function.  And indeed, we do.  This necessitates manually listing the packages to get them included in the function.  The code looks exactly as I think it should be written -- and how we want it to be written -- but I'm guessing something to do with the function code being separated into its own file tricked our module import capture detection logic.

**Comment by joeduffy** [comment]
Over in https://github.com/pulumi/pulumi-aws/issues/249, @rgwood said

> By the way, is requiring the compiled JS output like require('./bin/lambda'); the recommended approach for multi-file TypeScript projects like this? Not the end of the world but it feels a little awkward to reference compiled output instead of importing the Typescript class like import * as lambda from "./lambda"; (when I do that I still get a serialization error).

I have a feeling all of this is related.  It looks like

* when `import` is used, we see the import, attempt to serialize it, and fail
* when `require` is used, we somehow miss the import, and so don't even attempt to serialize it

Of course, in the latter case, we fail at runtime, because the runtime dependency is missing.

The reason we fail to serialize is because the `uuid` module uses native code:

```
              module './node_modules/uuid/index.js' which indirectly referenced
                function 'v4': v4.js(4,11): which captured
                  module './node_modules/uuid/lib/rng.js' which indirectly referenced
                    function 'nodeRNG': rng.js(6,33): which captured
                      module 'crypto' which indirectly referenced
                        function 'randomBytes': random.js(52,20): which referenced
                          function 'assertSize': random.js(33,19): which captured
                            'ERR_INVALID_ARG_TYPE', a function defined at
                              function 'NodeError': errors.js(156,15): which referenced
                                function 'getMessage': errors.js(208,19): which captured
                                  variable 'messages' which indirectly referenced
                                    function 'get': which could not be serialized because
                                      it was a native code function.
```

In summary, this is particularly unfortunate for a few reasons

1. why did we miss the capture if you resort to `require`? 
2. `crypto` is a built-in Node.js module, why did we even attempt to serialize it?
3. I think we should strongly reconsider reifying captures as `require`s at runtime

**Comment by CyrusNajmabadi** [comment]
> crypto is a built-in Node.js module, why did we even attempt to serialize it?

That is a good question we need to look into.  It's def in our list of whitelisted modules.  So something 'interesting' is going on and we def need to debug through to find out why this happened.

> why did we miss the capture if you resort to require?

I don't repro this myself.  If i ```require("uuid")``` or i ```await import("uuid")```, then both appropriate work at runtime.

> I think we should strongly reconsider reifying captures as requires at runtime

Definitely something we can consider again.  I will note that you were in the camp of strongly stating that modules should not be treated specially and we should just have a single consistent way to capture values, and we should use those for modules as well :)

--

Note: i think we're really in a rock and hard-place.  Reified modules have their own issues.  It def usually seems more pleasant initially (since people just write 'imports' exactly as normal), but the cracks leak through in other ways.  'inside require/import' is def less idiomatic, but avoids those issues and better reflects what would be going on anyways.  

I want more data to go off of before deciding if we need to do anything here.

**Comment by CyrusNajmabadi** [comment]
>>     crypto is a built-in Node.js module, why did we even attempt to serialize it?
> That is a good question we need to look into. It's def in our list of whitelisted modules. So something 'interesting' is going on and we def need to debug through to find out why this happened.

Actually, from looking at our code, we don't actually special case this at all in terms of capturing.  @lukehoban Do you remember if this was intentional.  From looking at our tests, it seems like we intended this, esp due to the confusion around what it means to use an externally imported module, inside a cloud-callback function.

I think a major thing we often hit was: what happens with a mutable/stateful module that is captured?  If you reify, then there is confusion because people don't see the state they mutated (which, is the pulumi-config case).  If you don't reify, then you cause serialization, which is only going to really work if the lib is designed for that.

So we settled on capture (and require/import on the inside if needed).

--

Note: its possible we could have a hybrid world.  One where we whitelist and *extremely* safe subset of modules to use.  i.e. things that are effectively stateless and which effectively feel "very safe" to reify.  But i'm not sure we need to do that, and i'd like to discuss things more before making any changes here that may not be necessary.

**Comment by CyrusNajmabadi** [comment]
@lukehoban   and i discussed this area quite a bit today.  We feel like there is likely room for improvements to the programming model here that will allow users to write code in the idiomatic style they expect, while also not opening up more cracks that lead to confusion.

The core thing we want to to solve is allowing users to be able to type ```import ... name ...``` at the top level, and then capture ```name``` inside 'cloud runtime functions' without hassle.  We think we can do that by re-introduce the concept of reifying captured modules as ```require```s inside captured functions. (note, there will be some tweaks here that will come lower)

Pros: 
1. Importing and using any *normal* JS module is trivial.  Users can use any idiomatic import style they want (i.e. ```outside import```, ```outside require```, ```inside await import```, ```inside require```), and it will all work.

Cons: 
1. Outside imports will 'resolve/execute' at deployment time, even if they are not needed at all during deployment.  This is likely not to actually be a problem.
2. Mutations to the module at deployment time will not be visible at runtime.  This may cause some confusion, but should be something we can easily explain to users.

Another concern we would have here *if we did nothing else* is that this would lead to a suboptimal experience around certain modules like ```@pulumi/pulumi```.  Namely, if we took the above approach **only**, then you would not be able to do ```new pulumi.Config("whatever")``` in your runtime code.  This is because that would translate into ```new require("pulumi").Config("whatever")```, which means your runtime code would never see the actual deployment time config values set whne we executed the pulumi-app.

To address that, we propose augment module serialization with the following capability:

1. Module objects can set a 'notAModule' bit on them, telling pulumi serialization to *not* treat a captured instance of as a module, and instead treat it like a normal value.  This will then cause us to have the same logic we have today, which means that 'Config' captures the current deployment time values for this object.

In essence this approach 'flips' how we treat modules today.  Today we say "all modules are values, and if you don't want that, don't even allow the module to go through closure serialization".  Instead, we'll have an approach that says "all modules are modules, and if you want a module to be captured by value, you mark it as such".

I personally think flipping things this way is appropriate.  In practice 99.99% of modules should not be treated as values.  So treating them as such, and basically failing 99.99% of hte time on them (since practically every module does something we cannot serialize) isn't a great default.

--

One concern we have here is that we have to be careful in terms of ensuring that the "reify as require" works properly for local-modules, and not just modules in the node_modules directory.

For example, if the user does: ```import * as x from "./x"```, we need to ensure that we are ```require```ing the right path.  Luke thinks this will actually work out thanks to how node module resolution works.  But we need to be careful we don't break this.

**Comment by lukehoban** [comment]
See also https://github.com/pulumi/pulumi-aws/issues/257 about an option to make the computation of packages to include in function zip uploads safer.  This may be even more needed if we adopt an approach like @CyrusNajmabadi suggests above.

**Comment by joeduffy** [comment]
I am a huge :+1::100: on this approach.
